internal/service/creditgrant.go

```diff
@@ -219,3 +221,546 @@ func (s *creditGrantService) GetCreditGrantsBySubscription(ctx context.Context,
+
+	// If successful and recurring, create next period application
+	if err == nil && grant.Cadence == types.CreditGrantCadenceRecurring {
+		nextErr := s.createNextPeriodApplication(ctx, grant, subscription, cga.PeriodEnd)
+		if nextErr != nil {
+			s.Logger.Errorw("failed to create next period application", "error", nextErr)
+		}
+	}
+
+	return err
+}
+
+// createNextPeriodApplication creates a new CGA entry with scheduled status for the next period
+func (s *creditGrantService) createNextPeriodApplication(ctx context.Context, grant *creditgrant.CreditGrant, subscription *subscription.Subscription, currentPeriodEnd time.Time) error {
+	// Calculate next period dates
+	nextPeriodStart, nextPeriodEnd, err := s.calculateNextPeriod(grant, subscription, currentPeriodEnd)
+	if err != nil {
+		s.Logger.Errorw("failed to calculate next period",
+			"grant_id", grant.ID,
+			"subscription_id", subscription.ID,
+			"current_period_end", currentPeriodEnd,
+			"error", err)
+		return err
+	}
+
+	// Create next period CGA
+	nextPeriodCGA := &domainCreditGrantApplication.CreditGrantApplication{
+		ID:                              types.GenerateUUIDWithPrefix(types.UUID_PREFIX_CREDIT_GRANT_APPLICATION),
+		CreditGrantID:                   grant.ID,
+		SubscriptionID:                  subscription.ID,
+		ScheduledFor:                    nextPeriodStart,
+		PeriodStart:                     nextPeriodStart,
+		PeriodEnd:                       nextPeriodEnd,
+		ApplicationStatus:               types.ApplicationStatusScheduled,
+		CreditsApplied:                  decimal.Zero,
+		Currency:                        subscription.Currency,
+		ApplicationReason:               "recurring_credit_grant_next_period",
+		SubscriptionStatusAtApplication: string(subscription.SubscriptionStatus),
+		RetryCount:                      0,
+		IdempotencyKey:                  s.generateIdempotencyKey(grant, subscription, nextPeriodStart, nextPeriodEnd),
+		EnvironmentID:                   types.GetEnvironmentID(ctx),
+		BaseModel:                       types.GetDefaultBaseModel(ctx),
+	}
+
+	err = s.CreditGrantApplicationRepo.Create(ctx, nextPeriodCGA)
+	if err != nil {
+		s.Logger.Errorw("failed to create next period CGA",
+			"next_period_start", nextPeriodStart,
+			"next_period_end", nextPeriodEnd,
+			"error", err)
+		return err
+	}
+
+	s.Logger.Infow("created next period credit grant application",
+		"grant_id", grant.ID,
+		"subscription_id", subscription.ID,
+		"next_period_start", nextPeriodStart,
+		"next_period_end", nextPeriodEnd,
+		"application_id", nextPeriodCGA.ID)
+
+	return nil
+}
+
+// calculateNextPeriod calculates the next credit grant period using simplified logic
+func (s *creditGrantService) calculateNextPeriod(grant *creditgrant.CreditGrant, subscription *subscription.Subscription, currentPeriodEnd time.Time) (time.Time, time.Time, error) {
+	nextPeriodStart := currentPeriodEnd
+
+	creditGrantPeriodConfig := map[types.CreditGrantPeriod]types.BillingPeriod{
```

Comment:
move this to types rather than inside a function so that this can be used at other places as well